File: Data_Structure/Insertion_Circular_LinkedList.c
#include<stdio.h>
#include<stdlib.h>

struct node{
    int data;
    struct node * next;
};

void linkedListTraversal(struct node *head)
{
    struct node *ptr = head;
    do
    {
        printf("%d ", ptr -> data);
        ptr = ptr -> next; 
    }while (ptr != head);
}

struct node* insertAtFirst(struct node* head, int data)
{
    struct node * ptr = (struct node *) malloc(sizeof(struct node));
    ptr -> data = data;
    struct node * p = head -> next;
    while(p -> next != head)
    {
        p = p -> next;
    }
    
    p -> next = ptr;
    ptr -> next = head;
    head = ptr;
    return head;
}

struct node* insertAtIndex(struct node* head, int data, int index)
{
    struct node* ptr = (struct node *)malloc(sizeof(struct node));
    ptr -> data = data;
    struct node* p = head;
    int i = 0;
    
    while(i != index-1)
    {
        p = p -> next;
        i++;
    }
    ptr -> next = p -> next;
    p -> next = ptr;
    return head;
}

struct node* insertAtEnd(struct node* head, int data)
{
    struct node * ptr = (struct node *) malloc(sizeof(struct node));
    ptr -> data = data;
    struct node * p = head -> next;
    while(p -> next != head)
    {
        p = p -> next;
    }
    
    p -> next = ptr;
    ptr -> next = head;
    return head;
}

struct node* insertNode(struct node* head, struct node* prevNode, int data)
{
    struct node * ptr = (struct node *) malloc(sizeof(struct node));
    ptr -> data = data;
    ptr -> next = prevNode -> next;
    prevNode -> next = ptr;
    return head;
}

int main(){
    struct node *head;
    struct node *second;
    struct node *third;
    struct node *fourth;

    // Allocate memory for nodes in the linked list in Heap
    head = (struct node *)malloc(sizeof(struct node));
    second = (struct node *)malloc(sizeof(struct node));
    third = (struct node *)malloc(sizeof(struct node));
    fourth = (struct node *)malloc(sizeof(struct node));

    // Link first and second nodes
    head -> data = 7;
    head -> next = second;

    // Link second and third nodes
    second -> data = 11;
    second -> next = third;

    // Link third and fourth nodes
    third -> data = 41;
    third -> next = fourth;

    // Terminate the list at the third node
    fourth -> data = 66;
    fourth -> next = head;

    printf("Linked list before insertion\n");
    linkedListTraversal(head);
    //head = insertAtFirst(head, 65);
    //head = insertAtIndex(head, 33, 2);
    //head = insertAtEnd(head, 46);
    head = insertNode(head, second, 57);
    printf("\nLinked list after insertion\n");
    linkedListTraversal(head);

    return 0;
}

File: Data_Structure/DS_Using_Python/Insertion_Circular_LinkedList.py
class Node:
	def __init__(self, data):
		self.data = data
		self.next = None
		
class CircularLinkedList:
	def __init__(self):
		self.head = None
		
	def printList(self):
		if self.head is None:
			print('List has no element')
			return
		else:
			temp = self.head
			while(temp):
				print(temp.data, end=' ')
				temp = temp.next
				if temp == self.head:
					break
	
	def linkedListLength(self):
				currNode = self.head
				length = 0
				while(currNode.next is not None):
					length += 1
					currNode = currNode.next
					if currNode == self.head:
						break
				return length
			
	#Insert a new node at the beginning	
	def push(self, data):
		newNode = Node(data)
		temp = self.head
		newNode.next = self.head
		if self.head is not None:
			while(temp.next != self.head):
				temp = temp.next
			temp.next = newNode
		else:
			newNode.next = newNode
		self.head = newNode
		
	#Inserting at the Beginning of the Linked List
	def insertAtFirst(self, val):
		if self.head is None:
		    print("List is empty")
		    return
		else:
		 	newNode = Node(val)
		 	temp = self.head
		 	while(temp.next != self.head):
		 		temp = temp.next
		 	temp.next = newNode
		 	newNode.next = self.head
		 	self.head = newNode
		 	
	#Inserting at the End of the Linked List
	def insertAtEnd(self, val):
		    if self.head is None:
		    	print("List is empty")
		    	return
		    else:
		    	newNode = Node(val)
		    	temp = self.head
		    	while(temp.next != self.head):
		    		temp = temp.next
		    	newNode.next = temp.next
		    	temp.next = newNode
	
	#Inserting in between two Data Nodes
	def insertAtIndex(self, val, index):
		if index < 0 or index > self.linkedListLength():
			print('\nThe index value is not correct')
			return
		else:
		 	newNode = Node(val)
		 	temp = self.head
		 	i = 0
		 	while(i != index-1):
		 		temp = temp.next
		 		i += 1
		 	newNode.next = temp.next
		 	temp.next = newNode
		 	
	#Searching an element
	def searchElement(self, key):
		temp = self.head
		q = self.head.next
		f = 0
		while(temp.next.next != self.head):
			if temp.data == key:
				f = 1
				break
			temp = temp.next
			q = q.next
			if q.next.data == key:
				f = 1
			if temp.next == self.head:
				break
		if f == 1:
		  print("\n{} is found".format(key));
		else:
		  print("\n{} is not found".format(key));

		 	
cllist = CircularLinkedList()

cllist.push(12) 
cllist.push(56) 
cllist.push(2) 
cllist.push(11) 

print('Linked list before insertion')
cllist.printList()

#cllist.insertAtFirst(62)
#print('\nLinked list after insertion')
#cllist.printList()

#cllist.insertAtEnd(57)
#print('\nLinked list after insertion')
#cllist.printList()

cllist.insertAtIndex(87, 3)
print('\nLinked list after insertion')
cllist.printList()

cllist.searchElement(162)